File: src/Butil/Bit.Butil/Scripts/serial.ts
var BitButil = (window as any).BitButil = (window as any).BitButil || {};

(function (butil: any) {
    // SerialPort objects carry the user's grant and stay on this side of the boundary; .NET works
    // through the Guid it minted for each one. See bluetooth.ts for the same shape.
    const _ports: { [id: string]: any } = {};
    // At most one read loop per port: the stream lock is exclusive, so a second reader would throw.
    const _readers: { [id: string]: { reader: any, subscriptionId: string, stopping: boolean } } = {};
    const _connectionListeners: { [id: string]: { connect: EventListener, disconnect: EventListener } } = {};
    // The line settings each open port was opened with. The API exposes no way to read them back and
    // no way to change them in place, so open() keeps them here to tell a redundant re-open from one
    // that is asking for different settings.
    const _openOptions: { [id: string]: any } = {};

    // Handle ids are minted here rather than by .NET, through the registry every device module
    // shares: the browser surfaces a port .NET has never seen, so there is nothing for it to key on
    // until the info comes back. Every entry point that hands one out hands out its id with it.
    const _registry = butil.utils.handleRegistry('sp', _ports);
    const idOf = _registry.idOf;

    butil.serial = {
        isSupported() { return !!(navigator as any).serial; },
        requestPort,
        getPorts,
        forget,
        release,
        open,
        close,
        isOpen,
        getInfo,
        write,
        writeText,
        startReading,
        stopReading,
        getSignals,
        setSignals,
        subscribeConnection,
        unsubscribeConnection
    };

    function serial() { return (navigator as any).serial; }

    function describe(id: string, port: any) {
        const info = typeof port.getInfo === 'function' ? (port.getInfo() ?? {}) : {};
        return {
            id,
            usbVendorId: info.usbVendorId ?? null,
            usbProductId: info.usbProductId ?? null,
            // readable/writable are non-null exactly while the port is open, which is the only
            // "is it open" signal the API gives.
            open: !!(port.readable || port.writable)
        };
    }

    async function requestPort(filters: any[]) {
        const api = serial();
        if (!api) return null;

        const requested = (filters ?? []).map((filter: any) => {
            const out: any = {};
            if (filter.usbVendorId != null) out.usbVendorId = filter.usbVendorId;
            if (filter.usbProductId != null) out.usbProductId = filter.usbProductId;
            return out;
        }).filter((filter: any) => Object.keys(filter).length > 0);

        let port: any;
        try {
            port = requested.length ? await api.requestPort({ filters: requested }) : await api.requestPort();
        } catch (e: any) {
            // NotFoundError is what dismissing the chooser looks like - "no port", not a failure.
            // Anything else is a mistake in the calling code and has to reach it.
            if (e?.name === 'NotFoundError') return null;
            throw e;
        }

        return describe(idOf(port), port);
    }

    async function getPorts() {
        const api = serial();
        if (!api) return [];
        try {
            const ports = await api.getPorts();
            return ports.map((port: any) => describe(idOf(port), port));
        } catch { return []; }
    }


    async function forget(id: string) {
        const port = _ports[id];
        if (!port || typeof port.forget !== 'function') return false;
        try { await port.forget(); release(id); return true; } catch { return false; }
    }

    function release(id: string) {
        const port = _registry.remove(id);
        if (!port) return;
        delete _openOptions[id];
        // Fire-and-forget: disposal must not wait on a port whose cable has already been pulled.
        // The read loop still holds the readable stream locked, though, and close() rejects until
        // it lets go - so the close is chained behind stopReading rather than raced with it.
        stopReading(id).then(() => port.close()).catch(() => { /* not open, or already gone */ });
    }

    async function open(id: string, options: any) {
        const port = _ports[id];
        if (!port) return false;

        const settings = {
            baudRate: options?.baudRate ?? 9600,
            dataBits: options?.dataBits ?? 8,
            stopBits: options?.stopBits ?? 1,
            parity: options?.parity ?? 'none',
            bufferSize: options?.bufferSize ?? 255,
            flowControl: options?.flowControl ?? 'none'
        };

        if (port.readable || port.writable) {
            // Re-opening with the settings already in force is the documented no-op. Different
            // settings are a real request, and the API cannot apply them to an open port - so the
            // port is cycled instead of reporting a success it did not deliver. That ends the read
            // loop exactly as close() does, and any subscription over it stops delivering.
            if (sameSettings(_openOptions[id], settings)) return true;
            await close(id);
        }

        await port.open(settings);
        _openOptions[id] = settings;
        return true;
    }

    function sameSettings(a: any, b: any) {
        if (!a) return false;
        return a.baudRate === b.baudRate && a.dataBits === b.dataBits && a.stopBits === b.stopBits
            && a.parity === b.parity && a.bufferSize === b.bufferSize && a.flowControl === b.flowControl;
    }

    async function close(id: string) {
        const port = _ports[id];
        if (!port) return;
        delete _openOptions[id];
        // The read loop holds the readable stream locked; close() throws until it lets go.
        await stopReading(id);
        try { await port.close(); } catch { /* not open, or already gone */ }
    }

    function isOpen(id: string) {
        const port = _ports[id];
        return !!(port && (port.readable || port.writable));
    }

    function getInfo(id: string) {
        const port = _ports[id];
        return port ? describe(id, port) : null;
    }

    async function writeBytes(id: string, data: Uint8Array) {
        const port = _ports[id];
        if (!port?.writable) return false;
        const writer = port.writable.getWriter();
        try { await writer.write(data); return true; }
        finally { writer.releaseLock(); }
    }

    function write(id: string, data: Uint8Array) { return writeBytes(id, data); }

    function writeText(id: string, text: string) {
        return writeBytes(id, new TextEncoder().encode(text));
    }

    // One chunk per dispatch, exactly as the device framed it - the API gives no message
    // boundaries, so re-assembling is the caller's decision and is left to .NET.
    function startReading(subscriptionId: string, dotNetRef: any, id: string) {
        const port = _ports[id];
        if (!port?.readable || _readers[id]) return false;

        const reader = port.readable.getReader();
        const entry = { reader, subscriptionId, stopping: false };
        _readers[id] = entry;

        (async () => {
            try {
                while (true) {
                    const { value, done } = await reader.read();
                    if (done) break;
                    if (value && value.length) butil.utils.dispatch(dotNetRef, 'InvokeSerialData', subscriptionId, value);
                }
            } catch (e: any) {
                // A cancel() during teardown rejects the pending read - not an error worth reporting.
                if (!entry.stopping) butil.utils.dispatch(dotNetRef, 'InvokeSerialError', subscriptionId, e?.message ?? String(e));
            } finally {
                try { reader.releaseLock(); } catch { /* stream already errored */ }
                if (_readers[id] === entry) delete _readers[id];
            }
        })();

        return true;
    }

    // A port carries at most one read loop, but that loop can end on its own - the device closed the
    // stream, or close() was called - and a later subscription can start a new one on the same port.
    // So a stop that names a subscription only stops the loop that subscription started: without the
    // check, disposing a stale subscription would cancel a reader it never began. Passing no
    // subscription id means "whatever is reading this port", which is what close() and release() want.
    async function stopReading(id: string, subscriptionId: string | null = null) {
        const entry = _readers[id];
        if (!entry) return;
        if (subscriptionId && entry.subscriptionId !== subscriptionId) return;
        entry.stopping = true;
        delete _readers[id];
        try { await entry.reader.cancel(); } catch { /* already cancelled */ }
        try { entry.reader.releaseLock(); } catch { /* released by the loop */ }
    }

    async function getSignals(id: string) {
        const port = _ports[id];
        if (!port || typeof port.getSignals !== 'function') return null;
        try {
            const signals = await port.getSignals();
            return {
                clearToSend: !!signals.clearToSend,
                dataCarrierDetect: !!signals.dataCarrierDetect,
                dataSetReady: !!signals.dataSetReady,
                ringIndicator: !!signals.ringIndicator
            };
        } catch { return null; }
    }

    async function setSignals(id: string, dataTerminalReady: boolean | null, requestToSend: boolean | null, brk: boolean | null) {
        const port = _ports[id];
        if (!port || typeof port.setSignals !== 'function') return false;
        const signals: any = {};
        // Omitted signals keep their current value; sending null would clear them.
        if (dataTerminalReady != null) signals.dataTerminalReady = dataTerminalReady;
        if (requestToSend != null) signals.requestToSend = requestToSend;
        if (brk != null) signals.break = brk;
        try { await port.setSignals(signals); return true; } catch { return false; }
    }

    // As in usb.ts: the event carries a port .NET may have no handle for, so one is minted here -
    // or the handle it already has is reused.
    function subscribeConnection(subscriptionId: string, dotNetRef: any) {
        const api = serial();
        if (!api) return false;

        const relay = (method: string) => ((event: any) => {
            const port = event.target ?? event.port;
            butil.utils.dispatch(dotNetRef, method, subscriptionId, describe(idOf(port), port));
        }) as EventListener;

        const connect = relay('InvokeSerialConnected');
        const disconnect = relay('InvokeSerialDisconnected');
        api.addEventListener('connect', connect);
        api.addEventListener('disconnect', disconnect);
        _connectionListeners[subscriptionId] = { connect, disconnect };
        return true;
    }

    function unsubscribeConnection(subscriptionId: string) {
        const entry = _connectionListeners[subscriptionId];
        if (!entry) return;
        delete _connectionListeners[subscriptionId];
        const api = serial();
        if (!api) return;
        api.removeEventListener('connect', entry.connect);
        api.removeEventListener('disconnect', entry.disconnect);
    }
}(BitButil));

File: src/Butil/Bit.Butil/Publics/Serial.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bit.Butil;

/// <summary>
/// Wraps the <see href="https://developer.mozilla.org/en-US/docs/Web/API/Web_Serial_API">Web Serial API</see>
/// (<c>navigator.serial</c>): pick a serial port, open it with the line settings the device
/// expects, and read and write bytes.
/// </summary>
/// <remarks>
/// Chromium only, and only over HTTPS. <see cref="RequestPort"/> opens the browser's port chooser
/// and so must run inside a user gesture.
/// </remarks>
[ButilService(typeof(Serial))]
public class Serial(IJSRuntime js) : IAsyncDisposable
{
    internal const string DataMethodName = nameof(InvokeSerialData);
    internal const string ErrorMethodName = nameof(InvokeSerialError);
    internal const string ConnectedMethodName = nameof(InvokeSerialConnected);
    internal const string DisconnectedMethodName = nameof(InvokeSerialDisconnected);

    // Keyed by subscription id, but carrying the port each one belongs to: serial.ts release() stops
    // that port's read loop, so the .NET closures behind it have to go at the same moment - otherwise
    // disposing one port leaks its handlers for as long as the service lives.
    private readonly ConcurrentDictionary<Guid, (string PortId, Action<byte[]> OnData, Action<string>? OnError)> _dataHandlers = new();
    private readonly ConcurrentDictionary<Guid, (Action<SerialPort>? OnConnected, Action<SerialPort>? OnDisconnected)> _connectionHandlers = new();

    // Every handle this service handed out, so a scope/circuit teardown closes the port even when
    // the caller never disposed it.
    private readonly ConcurrentDictionary<string, SerialPort> _ports = new();

    // Per-instance callback reference (see Keyboard): listeners are isolated per circuit / WASM app
    // and released on disposal - no static state, no cross-circuit leak.
    private DotNetObjectReference<Serial>? _dotNetRef;
    private DotNetObjectReference<Serial> DotNetRef => DotNetObjectReferenceHelper.GetOrCreate(ref _dotNetRef, this);

    /// <summary>True when the runtime exposes <c>navigator.serial</c>.</summary>
    /// <remarks>
    /// During prerender/SSR (no JS runtime) this returns <c>default</c> (e.g. <c>false</c>/<c>0</c>)
    /// rather than throwing, so the result can't be distinguished from a genuine value. If you
    /// branch on it, defer the read to <c>OnAfterRenderAsync</c>.
    /// </remarks>
    public ValueTask<bool> IsSupported() => js.Invoke<bool>("BitButil.serial.isSupported");

    /// <summary>
    /// Opens the browser's port chooser and returns the port the user picked, or null when they
    /// dismissed it. Must be called from a user gesture.
    /// </summary>
    /// <param name="filters">
    /// Which ports the chooser lists. Pass none to list every port - the usual choice, since a
    /// machine rarely has enough of them for filtering to help.
    /// </param>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(SerialPortInfo))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(SerialPortFilter))]
    public async ValueTask<SerialPort?> RequestPort(params SerialPortFilter[] filters)
    {
        var info = await js.Invoke<SerialPortInfo?>("BitButil.serial.requestPort", (object)filters);
        return info is null ? null : Track(info);
    }

    /// <summary>The ports this origin has already been granted, without showing a chooser.</summary>
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(SerialPortInfo))]
    public async ValueTask<SerialPort[]> GetPorts()
    {
        var infos = await js.Invoke<SerialPortInfo[]>("BitButil.serial.getPorts");
        return [.. infos.Select(Track)];
    }

    // One handle per browser-side id. serial.ts idOf() deliberately hands the same id back for the
    // same port, so a port surfaced again by GetPorts or by a connect event has to return the handle
    // already in play: a second wrapper over one JS registry entry would let disposing either of them
    // release the port out from under the other. Info stays the snapshot the first handle was created
    // from - it is documented as not updating on its own, and GetInfo() re-reads it.
    private SerialPort Track(SerialPortInfo info) => _ports.GetOrAdd(info.Id, _ => new SerialPort(this, js, info));

    // Called by a handle that is disposing itself, so the service stops holding it. The JS-side
    // release() stops that port's read loop, so the .NET closures behind it have to go with it -
    // otherwise disposing one port leaks the handlers of that port for as long as the service lives.
    // The registry removal is identity-checked so a handle that has already been replaced cannot
    // untrack its successor.
    internal void Forget(SerialPort port)
    {
        _ports.TryRemove(new KeyValuePair<string, SerialPort>(port.Id, port));

        foreach (var pair in _dataHandlers)
        {
            if (pair.Value.PortId == port.Id) _dataHandlers.TryRemove(pair.Key, out _);
        }
    }

    /// <summary>
    /// Invoked from JS for each chunk read from a port. Public + <see cref="JSInvokableAttribute"/>
    /// so it can be dispatched through the per-instance <see cref="DotNetObjectReference{T}"/>.
    /// </summary>
    [JSInvokable(DataMethodName)]
    public void InvokeSerialData(Guid id, byte[] data)
    {
        if (_dataHandlers.TryGetValue(id, out var handlers)) handlers.OnData.Invoke(data);
    }

    /// <summary>Invoked from JS when a read loop fails. See <see cref="InvokeSerialData"/>.</summary>
    [JSInvokable(ErrorMethodName)]
    public void InvokeSerialError(Guid id, string message)
    {
        if (_dataHandlers.TryGetValue(id, out var handlers)) handlers.OnError?.Invoke(message);
    }

    /// <summary>Invoked from JS on <c>navigator.serial</c>'s <c>connect</c>. See <see cref="InvokeSerialData"/>.</summary>
    [JSInvokable(ConnectedMethodName)]
    public void InvokeSerialConnected(Guid id, SerialPortInfo info)
    {
        if (_connectionHandlers.TryGetValue(id, out var handlers)) handlers.OnConnected?.Invoke(Track(info));
    }

    /// <summary>Invoked from JS on <c>disconnect</c>. See <see cref="InvokeSerialData"/>.</summary>
    [JSInvokable(DisconnectedMethodName)]
    public void InvokeSerialDisconnected(Guid id, SerialPortInfo info)
    {
        if (_connectionHandlers.TryGetValue(id, out var handlers)) handlers.OnDisconnected?.Invoke(Track(info));
    }

    [DynamicDependency(nameof(InvokeSerialData), typeof(Serial))]
    [DynamicDependency(nameof(InvokeSerialError), typeof(Serial))]
    internal async ValueTask<ButilSubscription> SubscribeData(string portId, Action<byte[]> onData, Action<string>? onError)
    {
        // The subscription id goes to stopReading along with the port id: the stream lock is
        // exclusive, so a port carries at most one read loop, but that loop can end on its own (the
        // device closed the stream, or Close() was called) and a later subscription can start a new
        // one. Stopping on the port alone would let a stale dispose cancel a reader it never started.
        return await ButilSubscriptionHelper.Register(_dataHandlers, (portId, onData, onError),
                                                      id => js.InvokeRegister("BitButil.serial.startReading", id, DotNetRef, portId),
                                                      id => js.InvokeVoid("BitButil.serial.stopReading", portId, id),
                                                      "Reading could not be started - the port is not open, or it is already being read.",
                                                      onError);
    }

    /// <summary>
    /// Watches ports appearing and disappearing - a USB serial adapter being plugged in or
    /// unplugged. Only ports this origin already has permission for raise these.
    /// </summary>
    /// <returns>A subscription - dispose it to detach the listeners.</returns>
    /// <exception cref="InvalidOperationException">The listeners were not attached - no Web Serial support.</exception>
    [DynamicDependency(nameof(InvokeSerialConnected), typeof(Serial))]
    [DynamicDependency(nameof(InvokeSerialDisconnected), typeof(Serial))]
    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(SerialPortInfo))]
    public async ValueTask<ButilSubscription> SubscribeConnection(Action<SerialPort>? onConnected = null,
                                                                  Action<SerialPort>? onDisconnected = null)
    {
        return await ButilSubscriptionHelper.Register(_connectionHandlers, (onConnected, onDisconnected),
                                                      id => js.InvokeRegister("BitButil.serial.subscribeConnection", id, DotNetRef),
                                                      id => js.InvokeVoid("BitButil.serial.unsubscribeConnection", id),
                                                      "The serial connection listeners could not be attached - the Web Serial API is not available.");
    }

    /// <summary>
    /// Closes every port this service handed out, detaches its listeners and releases the interop
    /// reference.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try
        {
            _dataHandlers.Clear();

            foreach (var id in _connectionHandlers.Keys.ToArray())
            {
                _connectionHandlers.TryRemove(id, out _);
                await js.InvokeVoid("BitButil.serial.unsubscribeConnection", id);
            }

            // Releasing a port stops its read loop, which is what actually detaches the data
            // handlers cleared above.
            foreach (var port in _ports.Values.ToArray())
            {
                await port.DisposeAsync();
            }
        }
        catch (Exception ex) when (ex.IsIgnorableDisposalException()) { } // teardown: circuit gone, cancelled, or already disposed
        finally
        {
            _dotNetRef?.Dispose();
            _dotNetRef = null;
        }

        GC.SuppressFinalize(this);
    }
}
